### ISSUE TITLE
Change text style of selected text

### ISSUE THREAD

**Issue by bdlukaa** [flutter/flutter#99231]
## Use case

I am trying to change the text style of the selected part of the text (`SelectableText`) to increase readability by changing the contrast of the colors. As you can see on the following image, the contrast between yellow and white is not pleasant to eye. I'd like to be able to change the white color to black only on the selected part of the text.

![contrast_example](https://user-images.githubusercontent.com/45696119/155881146-2aa5b7fe-3fe4-4dfe-9612-eb9f7af20056.png)

## Proposal

Be able to change the text style of only the selected part of the text, like the following:

![selection_example](https://user-images.githubusercontent.com/45696119/155881058-67ad096d-df65-4f3f-b070-bd6ab2133d62.png)

In the above image, the text was initially blue but it was changed to white when selected

**Comment by maheshmnj** [comment]
Hi @bdlukaa, Thanks for filing the issue.  Recently flutter introduced a new [Text Selection Theme](https://docs.flutter.dev/go/text-selection-theme) used to specify the properties of selected text in TextField and SelectableText widgets. 

```dart
ThemeData(
  textSelectionTheme: TextSelectionThemeData(
    cursorColor: Colors.red, 
    selectionColor: Colors.green,
    selectionHandleColor: Colors.blue,
  )
)
```


Please refer this doc for more info: https://docs.google.com/document/d/1sCe7Y_lhREljtn6m0Uu_EycbS82zRvWnNtkfEwFX9W0/edit
This is also documented on the official docs https://docs.flutter.dev/release/breaking-changes/text-selection-theme

<details>
<summary>minimal code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData.dark(),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Theme(
        data: ThemeData(
            textSelectionTheme: const TextSelectionThemeData(
          cursorColor: Colors.red,
          selectionColor: Colors.green,
          selectionHandleColor: Colors.blue,
        )),
        child: ListView(
          children: const [
            SelectableText(
              'Hello World ! This is a test of the SelectableText widget. ',
              style: TextStyle(fontSize: 20, color: Colors.white),
            )
          ],
        ),
      ),
    );
  }
}

```
</details>

#### output
 
<img width="585" alt="image" src="https://user-images.githubusercontent.com/31410839/155980790-a89e17e2-15c2-4818-a333-ead90cb518ca.png">

Closing as this is neither a bug nor a feature request.

**Comment by bdlukaa** [comment]
@maheshmnj I still believe this is a valid request because there isn't a property to change the text style (mainy the color) of the selected text. The image above isn't what I said.

For example, the initial text color is white, but I want to change only the selected text color to black, like the following image:
| Initially | After Selected |
| -------- | --------------- |
| ![Capturas 28_02_2022 10_21_28](https://user-images.githubusercontent.com/45696119/155990387-b3729991-3c0a-4e31-ba98-b3ddeaf64650.png) | ![](https://user-images.githubusercontent.com/45696119/155881058-67ad096d-df65-4f3f-b070-bd6ab2133d62.png) |

In the image above, the selected text color is white, but the rest of the color is like before.

Currently, there isn't a property to change the text style of only the selected text, only to change the background color, cursor color and the handles color.

**Comment by maheshmnj** [comment]
@bdlukaa sorry I should have mentioned it the `style` property of `SelectableText` should be used to change the style of the text.

This is the snippet for the above code sample.

```dart
SelectableText(
              'Hello World ! This is a test of the SelectableText widget. ',
              style: TextStyle(fontSize: 20, color: Colors.red),
            )
```

<img width="585" alt="image" src="https://user-images.githubusercontent.com/31410839/155991311-4a342b0f-902a-41e3-a2f4-7c8d83835bbf.png">

**Comment by bdlukaa** [comment]
@maheshmnj but there isn't a way to change the style of the selected part of the text. `style` on `SelectableText` changes the style whole text

**Comment by maheshmnj** [comment]
Oh, I see gotcha, That looks like a valid feature request. Thanks for the clarification.

SelectableText should support changing the style of the selectedText, Currently `SelectableText.style` property applies to both   Selected and unselected Text.

**Comment by maheshmnj** [comment]
Oh, I see gotcha, That looks like a valid feature request. Thanks for the clarification.

SelectableText should support changing the style of the selectedText, Currently `SelectableText.style` property applies to both   Selected and unselected Text.

**Comment by HansMuller** [comment]
CC @justinmc

**Comment by chunhtai** [comment]
Looks like, we can add something like selectionTextStyle to TextSelectionThemeData

**Comment by reiko-dev** [comment]
+1
Which workaround could be made on this?

**Comment by justinmc** [comment]
@reiko-dev You could work around this by overriding [TextEditingController.buildTextSpan](https://master-api.flutter.dev/flutter/widgets/EditableTextState/buildTextSpan.html).

I agree with @chunhtai about the full solution: Add selectionTextStyle to TextSelectionThemeData, then modify TextEditingController.buildTextSpan to use that style for the selected part of the text.

CC @Renzo-Olivares @camsim99 I think you guys have both dealt with trying to merge TextSpan trees recently.  Do we have any easy generic way of doing that or does this need something custom separate from deltas and spell check?

**Comment by camsim99** [comment]
> CC @Renzo-Olivares @camsim99 I think you guys have both dealt with trying to merge TextSpan trees recently. Do we have any easy generic way of doing that or does this need something custom separate from deltas and spell check?

Yes, I currently am doing that for spell check, since text needs to be styled depending on whether or not it is misspelled. My solution for that is currently tied to spell check specific logic, but I would be interested in generalizing it to address this issue because it is certainly achievable and overall better practice!

**Comment by reiko-dev** [comment]
> @reiko-dev You could work around this by overriding [TextEditingController.buildTextSpan](https://master-api.flutter.dev/flutter/widgets/EditableTextState/buildTextSpan.html).
> 
> I agree with @chunhtai about the full solution: Add selectionTextStyle to TextSelectionThemeData, then modify TextEditingController.buildTextSpan to use that style for the selected part of the text.
> 
> CC @Renzo-Olivares @camsim99 I think you guys have both dealt with trying to merge TextSpan trees recently. Do we have any easy generic way of doing that or does this need something custom separate from deltas and spell check?

Thanks @justinmc i'll try that

**Comment by Renzo-Olivares** [comment]
@justinmc I think right now because the composing region is not active when there is an active selection and vice versa, this may be a simple merge. 

Is what we do for composingStyle.

```dart
    final TextStyle composingStyle = style?.merge(const TextStyle(decoration: TextDecoration.underline))
        ?? const TextStyle(decoration: TextDecoration.underline);
    return TextSpan(
      style: style,
      children: <TextSpan>[
        TextSpan(text: value.composing.textBefore(value.text)),
        TextSpan(
          style: composingStyle,
          text: value.composing.textInside(value.text),
        ),
        TextSpan(text: value.composing.textAfter(value.text)),
      ],
    );
```

For selectionStyle something similar can be done.
```dart
    final TextStyle selectedTextStyle = style?.merge(selectionTextStyle)
        ?? selectionTextStyle;

    return TextSpan(
      style: style,
      children: <TextSpan>[
        TextSpan(text: value.selection.textBefore(value.text)),
        TextSpan(
          style: composingStyle,
          text: value.selection.textInside(value.text),
        ),
        TextSpan(text: value.selection.textAfter(value.text)),
      ],
    );
```

With some case to do either merge depending on if the composing region is active or not.

**Comment by juniorbomb** [comment]
any update guys?

**Comment by TENX-S** [comment]
Any updates on this?

**Comment by TENX-S** [comment]
Any updates on this?

**Comment by justinmc** [comment]
@Renzo-Olivares @hangyujin Any possibility of changing selection style with your work on SelectionArea and SelectableText?

**Comment by Vextrov** [comment]
I support this. In a **Text** widget there is a property called "_selectionColor_" which doesn't even do anything because you can't select the text. In **SelectableText** widget however, there is no "selectionColor". Specifying the color in _Theme.of(context).textSelectionTheme. selectionColor_ does not help in making it dynamic.

I think both of the following properties should be added to the **SelectableText** widget to allow dynamic and convenience changes:
1. Color? _selectionColor_
2. Color? _selectedTextColor_

**Comment by Renzo-Olivares** [comment]
@Vextrov you can make a `Text` widget selectable by wrapping it with a `SelectionArea` widget, then you can utilize the `selectionColor` property. WRT to `SelectableText` you can wrap it with a `DefaultSelectionStyle` widget and set the `selectionColor` that way. This is the color the selection highlight is painted in, which is different than the text style of the selected text that this issue is requesting to configure.

@justinmc I don't think we had any plans for this, but it sounds like a reasonable feature to add. Right now `SelectionArea` doesn't directly affect any of its `Selectable`s `TextStyle`s, that is up to the selectable widget itself at the moment.

**Comment by Isakdl** [comment]
This issue seems to be collecting dust, but I think this is a very reasonable feature and would push Flutters viability on web!

**Comment by rafuck** [comment]
Any updates on this feature guys? ,)
